### PR TITLE
fix(docker): run in docker, and build scripts for docker

### PR DESCRIPTION
**Pull request by egidijus**
no reason to grant excessive permissions to your docker container, this runs headless, no errors from chromium or puppetteer.

### Description

This also prints the config for puppetteer, makes it easier to troubleshoot:
```
streetmerchant-shop-bot    | [8:08:40 PM] warn :: ℹ printing puppeteer configs:: {
streetmerchant-shop-bot    |   "0": "--disable-dev-shm-usage",
streetmerchant-shop-bot    |   "1": "--no-sandbox",
streetmerchant-shop-bot    |   "2": "--disable-setuid-sandbox",
streetmerchant-shop-bot    |   "3": "--headless",
streetmerchant-shop-bot    |   "4": "--disable-gpu"
streetmerchant-shop-bot    | }
```

no more this error when running in docker
```
streetmerchant-shop-bot    | }
streetmerchant-shop-bot    | [8:05:39 PM] error :: ✖ something bad happened, resetting streetmerchant in 5 seconds Failed to launch the browser process!
streetmerchant-shop-bot    | Failed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Operation not permitted
streetmerchant-shop-bot    | [1208/200539.568829:FATAL:zygote_host_impl_linux.cc(191)] Check failed: ReceiveFixedMessage(fds[0], kZygoteBootMessage, sizeof(kZygoteBootMessage), &boot_pid). 
streetmerchant-shop-bot    | Received signal 6
streetmerchant-shop-bot    |   r8: 00007ddd4844  r9: 00007f3d3f6f40 r10: 0000000000000008 r11: 0000000000000246
streetmerchant-shop-bot    |  r12: 00007ffd80 r13: 00007ffddbc r14: 00007f30c4ad2170 r15: 00000000000000a0
streetmerchant-shop-bot    |   di: 0000000000000002  si: 00007dd0  bp: 00007ddd4710  bx: 0000000000000000
streetmerchant-shop-bot    |   dx: 0000000000000000  ax: 0000000000000000  cx: 00007f30cd8a71  sp: 00007ffc55d708
streetmerchant-shop-bot    |   ip: 00007d8a71 efl: 0000000000000246 cgf: 002b000000000033 erf: 0000000000000000
streetmerchant-shop-bot    |  trp: 0000000000000000 msk: 0000000000000000 cr2: 0000000000000000
streetmerchant-shop-bot    | [end of stack trace]
streetmerchant-shop-bot    | Calling _exit(1). Core file will not be generated.
streetmerchant-shop-bot    | 
streetmerchant-shop-bot    | 
streetmerchant-shop-bot    | TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md
streetmerchant-shop-bot    |  {
streetmerchant-shop-bot    |   "stack": "Error: Failed to launch the browser process!\nFailed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Operation not permitted\n[1208/200539d8829:FATAL:zygote_host_impl_linux.cc(191)] Check failed: ReceiveFixedMessage(fds[0], kZygoteBootMessage, sizeof(kZygoteBootMessage), &boot_pid). \nReceived signal 6\n  r8: 00007d52d4844  r9: 00007d63f6f40 r10: 0000000000000008 r11: 0000000000000246\n r12: 0000d80 r13: 00007d4dbc r14: 0000d170 r15: 00000000000000a0\n  di: 0000000000000002  si: 0000d4710  bp: 00007d4710  bx: 0000000000000000\n  dx: 0000000000000000  ax: 0000000000000000  cx: 00007d71  sp: 0000d8\n  ip: 0000d8a71 efl: 0000000000000246 cgf: 002b000000000033 erf: 0000000000000000\n trp: 0000000000000000 msk: 0000000000000000 cr2: 0000000000000000\n[end of stack trace]\nCalling _exit(1). Core file will not be generated.\n\n\nTROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md\n\n    at onClose (/app/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserRunner.js:193:20)\n    at ChildProcess.<anonymous> (/app/node_modules/puppeteer/lib/cjs/puppeteer/node/BrowserRunner.js:184:79)\n    at ChildProcess.emit (events.js:327:22)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)"
streetmerchant-shop-bot    | }
```

### Testing
Running `make` will do everything for you

### New dependencies
* Make
* docker-compose

### CODE REVIEW

**Comment by jef** [comment]
Love this! Does docker-compose come with installs on Windows? I know for macOS and Linux you need to install separately, so it would need to be a prerequisite.

**Comment by joshstubble** [comment]
Docker Compose does come installed in the docker desktop app on windows

**Comment by jef** [comment]
On second thought, I'm a little hesitant on the `docker-compose`. Not that it's bad to have or even gives a quality of life, but I think the `Makefile` alone is sufficient. And that may even be something I don't necessarily want to maintain either, along with `docker-compose`.

I'd also like to add that it's somewhat verbose given we can use `env-file` with `docker-run`. The main benefit of using `docker-compose` is utilizing multiple containers and/or advanced volumes and networks -- neither of which streetmerchant has. That's my knee jerk reaction. 

That being said, I really like the changes and it seems to be working great on my end. I appreciate your contribution and would like to hear your thoughts!

Thanks.

File: Dockerfile [hunk 1]
```diff
@@ -23,6 +23,8 @@ ENV PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser \
 	DOCKER=true
 
 RUN addgroup -S appuser && adduser -S -g appuser appuser \
+	&& addgroup appuser audio \
+	&& addgroup appuser video \
```
[the remark this answers is not in the document]
Comment (egidijus) [reply Dockerfile:27]: You're right, this is not needed.
Testing options bit by bit, will update the PR with minimal changes to get it working.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+      context: .
+      dockerfile: Dockerfile
+    container_name: streetmerchant-shop-bot
+    hostname: streetmerchant-shop-bot
+    # cap_add:
+    #   - CAP_SYS_ADMIN
+    environment:
+      - "ENV=dev"
+    volumes:
+      - ".env:/app/.env:ro"
```
[the remark this answers is not in the document]
Comment (egidijus) [reply docker-compose.yml:14]:
```
ENV=dev
```
It's a pattern I adopted in the industry.
If a thing is running anywhere, emit some kind of environment name. Useful if you have many things running and all the containers register in some kind of discovery service via consul +  registrator (https://github.com/gliderlabs/registrator)



```
volumes
```
Least privilege, no clobbering of files. My assumption is that an env file like this is easier to manipulate than to add another bunch of vars to environment in the docker compose.

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    ports:
+      - 7997:7997
```
[the remark this answers is not in the document]
Comment (egidijus) [reply docker-compose.yml:16]: Noted, PR update coming.

**Comment by egidijus** [comment]
> On second thought, I'm a little hesitant on the `docker-compose`. Not that it's bad to have or even gives a quality of life, but I think the `Makefile` alone is sufficient. And that may even be something I don't necessarily want to maintain either, along with `docker-compose`.
> 
> I'd also like to add that it's somewhat verbose given we can use `env-file` with `docker-run`. The main benefit of using `docker-compose` is utilizing multiple containers and/or advanced volumes and networks -- neither of which streetmerchant has. That's my knee jerk reaction.
> 
> That being said, I really like the changes and it seems to be working great on my end. I appreciate your contribution and would like to hear your thoughts!
> 
> Thanks.

🤷‍♀️ 
In my experience if you are working locally, even if it's a single container `docker-compose build/up` > `plain docker run` .

I could change the `docker-compose.yml` to v 2.0, you could use the docker-compose.yml to spin things up on portainer, some people run that on their NAS devices.

Obviously I like docker-compose and Makefiles, because it makes it super easy to get started without even reading any docs 😁

I see makefiles and docker-compose as build tools, once you get them going, they kinda work without much attention (caveat: if you have complex systems, your build tools will get complex, this is not complex).

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	# docker-compose --env-file .env up -d
+
+
+stop:
+	docker-compose stop
+
+clean:
+	docker-compose kill
+	docker-compose down
+	docker-compose rm -f
```
Comment (jef) [inline Makefile:22]:
```suggestion
.DEFAULT_GOAL := run

.PHONY: build
build:
	docker-compose build streetmerchant

.PHONY: run
run:
	docker-compose up

.PHONY: run-detached
run-detached:
	docker-compose up -d

.PHONY: stop
stop:
	docker-compose down

```

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+version: '3.0'
```
Comment (jef) [inline docker-compose.yml:1]:
```suggestion
version: '3.8'
```

Let's pin to the latest version.

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+services:
+  streetmerchant-shop-bot:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    container_name: streetmerchant-shop-bot
+    hostname: streetmerchant-shop-bot
```
Comment (jef) [inline docker-compose.yml:8]:
```suggestion
  streetmerchant:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: streetmerchant
```

I'd prefer we keep the name the same as the application. I also don't think it's necessary to have a `hostname`.

File: docker-compose.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jef) [reply docker-compose.yml:12]: > It's a pattern I adopted in the industry.
> If a thing is running anywhere, emit some kind of environment name. Useful if you have many things running and all the containers register in some kind of discovery service via consul + registrator (gliderlabs/registrator)

As much as I can understand where you're coming from, I haven't seen this pattern much and prefer not to use it here.

As for the volumes, I think this is a great idea! Although, `env_file: .env` is just as efficient and specific.

```suggestion
    env_file: .env
```

**Review by jef — COMMENTED** [review]
Okay, I see where you're coming from. If you don't mind though, I'll make a couple suggestions and you let me know what you think.

I really appreciate your contribution and your time for the input and feedback!

File: docker-compose.yml (hunk 5 above)
Comment (egidijus) [reply docker-compose.yml:1]: I haven't tested if this docker-compose is compatible with v3.8 compose file.
these versions are for the `docker-compose.yml` file, across the versions some add more features at the cost of mandatory complexity.

We can stick with v3.0 of docker-compose file until the end of days (caveat: if we do not need any functionality from greater versions, and do not mind slight change of syntax).

https://docs.docker.com/compose/compose-file/

**Review by jef — APPROVED** [review]
Amazing work!!! I really appreciate your feedback and the time you took to implement this.

**Comment by rchenzheng** [comment]
Thanks @jef @egidijus

**Comment by TDCroPower** [comment]
the Docker image was unfortunately only built for amd64 and therefore does not work on my Helios64 with rockchip64  (arm64/v8) architecture.
Do you have a tip how I can rewrite the Dockerfile so that I can build it on the Helios64 as well?

here the output from Helios64...
```
root@helios64:/opt/docker/streetmerchant# docker run -it --rm \
>   --env-file ./dotenv \
>   ghcr.io/jef/streetmerchant:nightly
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
standard_init_linux.go:219: exec user process caused: exec format error
root@helios64:/opt/docker/streetmerchant# uname -a
Linux helios64 5.9.14-rockchip64 #20.11.4 SMP PREEMPT Tue Dec 15 08:52:20 CET 2020 aarch64 GNU/Linux
root@helios64:/opt/docker/streetmerchant#
```

**Comment by kiowadriver** [comment]
I am able to run fine without the SYS_ADMIN parameter (less of a security risk).  

I have included some build instructions in my fork which i am keeping up to date with the main branch with some minor tweaks for unraid usage:
https://github.com/kiowadriver/streetmerchant/blob/main/README.md

**Comment by rchenzheng** [comment]
> I am able to run fine without the SYS_ADMIN parameter (less of a security risk).
> 
> I have included some build instructions in my fork which i am keeping up to date with the main branch with some minor tweaks for unraid usage:
> https://github.com/kiowadriver/streetmerchant/blob/main/README.md

It has already been merged

/close

**Comment by TDCroPower** [comment]
i first had the problem on my helios with aarch64 and was able to work around it.
Now I wanted to switch to the OrangePi to run the script on the side, but there is again the problem with the architecture because this uses armv7l and there is no chromium in the package manager...

aarch64...
https://pkgs.alpinelinux.org/packages?name=chromium&branch=edge&arch=aarch64

armv7l...
https://pkgs.alpinelinux.org/packages?name=chromium&branch=edge&arch=armv7

should I rather create a new issue for this or can we discuss it further here?

**Comment by jef** [comment]
If you're having problems with install the Chromium browser, it's because of the package manager used in alpine is `apk` and debian uses `apt`. You'll need to replace those lines.

I prefer that we keep this image using alpine, as it should work on most systems.

If you'd like make this work for your ARM architecture, I think you should change the image to use `arm64v8/node:14.15.0-alpine3.12`.

**Comment by TDCroPower** [comment]
@jef the problem no longer exists for the Helios65 arm64v8, there it worked in the end, because for this there is the Chromium package at alpine.
However, there is no Chromium for armv7l at alpine, see link above.
Would need an alternative image or another variant to install Chromium in the image.

I am currently testing to rewrite the Dockerfile so that I work with buster-slim, since there should be the Chromium package on all architectures.

**Pull request merged**